tools/codegen/api/dispatcher.py

```diff
@@ -137,7 +137,22 @@ def cppargument_exprs(
     else:
         assert_never(a)
 
-def cpparguments_exprs(args: Sequence[CppArgumentPack]) -> Sequence[DispatcherExpr]:
+def cpparguments_exprs(func: FunctionSchema, * , method: bool, api_is_faithful: bool) -> Sequence[DispatcherExpr]:
+    dispatcher_calling_convention_is_faithful = local.use_c10_dispatcher().dispatcher_uses_new_style()
+    arguments = cpp.group_arguments(func, method=method, faithful=dispatcher_calling_convention_is_faithful)
+
+    if api_is_faithful:
+        argument_packs = tuple(
+            cpp.argument_faithful(a) for a in arguments
+        )
+    else:
+        argument_packs = tuple(
+            cpp.argument(a) for a in arguments
+        )
```

Comment:
You do grouping (with faithfulness) and then mapping between argument and `argument_faithful` here, but then you replicate the logic again at https://github.com/pytorch/pytorch/pull/47712/files#diff-2cccf17f15442a35105f349b3bebaf199294d8c3380767b73117031589832634R191 This seems a bit smelly to me.

Reply:
the reason is that one thing branches based on if the dispatcher calling convention is faithful and the other thing branches based on if the API is faithful. Those are different things, we're actually handling 2x2=4 scenarios here.